grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancer.java

```diff
@@ -476,24 +484,74 @@ public PickResult pickSubchannel(PickSubchannelArgs args) {
   }
 
   @VisibleForTesting
+  static final class RoundRobinEntry {
+    final PickResult result;
+    @Nullable
+    final String token;
+
+    /**
+     * A non-drop result.
+     */
+    RoundRobinEntry(Subchannel subchannel, String token) {
+      this.result = PickResult.withSubchannel(subchannel);
+      this.token = token;
+    }
+
+    /**
+     * A drop result.
+     */
+    RoundRobinEntry(Status dropStatus) {
+      this.result = PickResult.withError(dropStatus);
+      this.token = null;
+    }
+
+    void updateHeaders(Metadata headers) {
+      if (token != null) {
+        headers.discardAll(TOKEN_KEY);
+        headers.put(TOKEN_KEY, token);
+      }
+    }
+
+    @Override
+    public String toString() {
+      return "[result=" + result + ", token=" + token + "]";
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(result, token);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      if (!(other instanceof RoundRobinEntry)) {
+        return false;
+      }
+      RoundRobinEntry that = (RoundRobinEntry) other;
+      return Objects.equal(result, that.result) && Objects.equal(token, that.token);
+    }
+  }
+
+  @VisibleForTesting
   static final class RoundRobinPicker extends SubchannelPicker {
-    final List<PickResult> list;
+    final List<RoundRobinEntry> list;
     int index;
 
-    RoundRobinPicker(List<PickResult> resultList) {
+    RoundRobinPicker(List<RoundRobinEntry> resultList) {
       checkArgument(!resultList.isEmpty(), "resultList is empty");
       list = resultList;
     }
 
     @Override
     public PickResult pickSubchannel(PickSubchannelArgs args) {
       synchronized (list) {
-        PickResult result = list.get(index);
+        RoundRobinEntry result = list.get(index);
         index++;
         if (index == list.size()) {
           index = 0;
         }
-        return result;
+        result.updateHeaders(args.getHeaders());
```

Comment:
this is counter-intuitive: `public PickResult pickSubchannel(PickSubchannelArgs args)` will produce a PickResult which does not depend on the args at all, but will mutate the args.